File: src/models/ShelterCategory.ts
import Shape from "./Shape";
import ShelterImage from "./ShelterImage";

class ShelterCategory {
    id?: number;
    shelterCategoryName?: string;
    width?: number;
    height?: number;
    length?: number;
    diameter?: number;
    waterVolume?: number;
    waterFiltrationSystem?: string;
    description?: string;
    status?: string;
    createdDate?: Date;
    createdBy?: string;
    updatedDate?: Date;
    updatedBy?: string;
    shape?: Shape;
    shelterImages?: ShelterImage[]; 

    constructor(
        id?: number,
        shelterCategoryName?: string,
        width?: number,
        height?: number,
        length?: number,
        diameter?: number,
        waterVolume?: number,
        waterFiltrationSystem?: string,
        description?: string,
        status?: string,
        createdDate?: Date,
        createdBy?: string,
        updatedDate?: Date,
        updatedBy?: string,
        shape?: Shape, 
        shelterImages?: ShelterImage[], 
    ) {
        this.id = id;
        this.shelterCategoryName = shelterCategoryName;
        this.width = width;
        this.height = height;
        this.length = length;
        this.diameter = diameter;
        this.waterVolume = waterVolume;
        this.waterFiltrationSystem = waterFiltrationSystem;
        this.description = description; 
        this.status = status;
        this.createdDate = createdDate;
        this.createdBy = createdBy;
        this.updatedDate = updatedDate;
        this.updatedBy = updatedBy;
        this.shape = shape; 
        this.shelterImages = shelterImages; 
    }
}

export default ShelterCategory;

File: BackEnd/FA24_SE1854_SWP391_G8_FengShuiKoiConsultingSystem/demo/src/main/java/com/fengshuisystem/demo/entity/ShelterCategory.java
package com.fengshuisystem.demo.entity;

import com.fengshuisystem.demo.entity.enums.Status;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.Nationalized;

import java.time.Instant;
import java.util.LinkedHashSet;
import java.util.Set;

@Getter
@Setter
@Entity
public class ShelterCategory {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "shelter_category_id", nullable = false)
    private Integer id;

    @Size(max = 255)
    @NotNull
    @Column(name = "shelter_category_name", nullable = false)
    private String shelterCategoryName;

    @NotNull
    @ManyToOne(cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH, CascadeType.DETACH}, optional = false)
    @JoinColumn(name = "shape_id", nullable = false)
    private Shape shape;

    @Column(name = "width")
    private Double width;

    @Column(name = "height")
    private Double height;

    @Column(name = "length")
    private Double length;

    @Column(name = "diameter")
    private Double diameter;

    @Column(name = "water_volume")
    private Double waterVolume;

    @Size(max = 255)
    @Nationalized
    @Column(name = "water_filtration_system")
    private String waterFiltrationSystem;

    @Size(max = 1000)
    @Nationalized
    @Column(name = "description", length = 1000)
    private String description;

    @NotNull
    @Column(name = "status")
    @Enumerated(EnumType.STRING)
    private Status status = Status.ACTIVE;

    @NotNull
    @Column(name = "created_date", nullable = false)
    private Instant createdDate;

    @Size(max = 255)
    @NotNull
    @Nationalized
    @Column(name = "created_by", nullable = false)
    private String createdBy;

    @Column(name = "updated_date")
    private Instant updatedDate;

    @Size(max = 255)
    @Nationalized
    @Column(name = "updated_by")
    private String updatedBy;

    // Chuyển từ OneToMany sang ManyToMany với ConsultationRequestDetail
    @ManyToMany(fetch = FetchType.LAZY, cascade = CascadeType.MERGE)
    @JoinTable(
            name = "consultation_shelter_category",
            joinColumns = @JoinColumn(name = "shelter_category_id"),
            inverseJoinColumns = @JoinColumn(name = "consultation_request_detail_id")
    )
    private Set<ConsultationRequestDetail> consultationRequestDetails = new LinkedHashSet<>();

    @OneToMany(mappedBy = "shelterCategory", cascade = {CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH, CascadeType.DETACH})
    private Set<ConsultationShelter> consultationShelters = new LinkedHashSet<>();

    @OneToMany(mappedBy = "shelterCategory", cascade = {CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REFRESH, CascadeType.DETACH})
    private Set<ShelterImage> shelterImages = new LinkedHashSet<>();
}
